google-cloud-firestore/lib/google/cloud/firestore/collection_reference_list.rb

```diff
@@ -0,0 +1,137 @@
+          self.class.from_grpc grpc, @client, @parent, max: @max
+        end
+
+        ##
+        # Retrieves remaining results by repeatedly invoking {#next} until {#next?} returns `false`. Calls the given
+        # block once for each result, which is passed as the argument to the block.
+        #
+        # An Enumerator is returned if no block is given.
+        #
+        # This method will make repeated API calls until all remaining results are retrieved. (Unlike `#each`, for
+        # example, which merely iterates over the results returned by a single API call.) Use with caution.
+        #
+        # @param [Integer] request_limit The upper limit of API requests to make to load all collection references.
+        #   Default is no limit.
+        # @yield [collection_reference] The block for accessing each collection_reference.
+        # @yieldparam [CollectionReference] collection_reference The collection reference object.
+        #
+        # @return [Enumerator]
+        #
+        # @example Iterating each collection reference by passing a block:
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   firestore.cols do |collection_reference|
+        #     puts collection_reference.collection_id
+        #   end
+        #
+        # @example Using the enumerator by not passing a block:
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   all_collection_ids = firestore.cols.map do |collection_reference|
+        #     collection_reference.collection_id
+        #   end
+        #
+        def all request_limit: nil
+          request_limit = request_limit.to_i if request_limit
+          unless block_given?
+            return enum_for :all, request_limit: request_limit
+          end
+          results = self
+          loop do
+            results.each { |r| yield r }
+            if request_limit
+              request_limit -= 1
+              break if request_limit < 0
+            end
+            break unless results.next?
+            results = results.next
+          end
+        end
+
+        ##
+        # @private New CollectionReference::List from a `Google::Cloud::Firestore::V1::ListCollectionIdsResponse`
+        # object.
+        def self.from_grpc grpc, client, parent, max: nil
+          raise ArgumentError, "parent is required" unless parent
+          cols = CollectionReferenceList.new(Array(grpc.collection_ids).map do |collection_id|
+            CollectionReference.from_path "#{parent}/#{collection_id}", client
+          end)
+          token = grpc.next_page_token
+          token = nil if token == "".freeze
+          cols.instance_variable_set :@token, token
+          cols.instance_variable_set :@client, client
+          cols.instance_variable_set :@parent, parent
+          cols.instance_variable_set :@max, max
```

Comment:
Also curious why we're doing this, rather than passing these values into the constructor and setting the instance variables there. (We don't have to change it now, and I understand that DocumentReference::List is set up the same way. Just seems unusual and I wanted to understand if there's a reason for it.)

Reply:
I asked about the reason for this pattern back in 2015-6, but I can't remember Mike's answer. I would be more than happy to change this to plain old variable inits in the constructor.

However, the pattern is not at all unusual at this point in google-cloud-ruby. Here's a search for `instance_variable_set :@token`:

```
27 results - 26 files

google-cloud-ruby • google-cloud-bigquery/lib/google/cloud/bigquery/dataset/list.rb:
  142:             datasets.instance_variable_set :@token,   gapi_list.next_page_token

google-cloud-ruby • google-cloud-bigquery/lib/google/cloud/bigquery/job/list.rb:
  144:             jobs.instance_variable_set :@token,    gapi_list.next_page_token

google-cloud-ruby • google-cloud-bigquery/lib/google/cloud/bigquery/model/list.rb:
  146:             models.instance_variable_set :@token,      gapi_list[:nextPageToken]

google-cloud-ruby • google-cloud-bigquery/lib/google/cloud/bigquery/project/list.rb:
  143:             projects.instance_variable_set :@token,   gapi_list.next_page_token

google-cloud-ruby • google-cloud-bigquery/lib/google/cloud/bigquery/routine/list.rb:
  146:             routines.instance_variable_set :@token,      gapi_list.next_page_token

google-cloud-ruby • google-cloud-bigquery/lib/google/cloud/bigquery/table/list.rb:
  152:             tables.instance_variable_set :@token,      gapi_list.next_page_token

google-cloud-ruby • google-cloud-dns/lib/google/cloud/dns/change/list.rb:
  153:             changes.instance_variable_set :@token, gapi.next_page_token

google-cloud-ruby • google-cloud-dns/lib/google/cloud/dns/record/list.rb:
  153:             records.instance_variable_set :@token, gapi.next_page_token

google-cloud-ruby • google-cloud-dns/lib/google/cloud/dns/zone/list.rb:
  148:             zones.instance_variable_set :@token,   gapi.next_page_token

google-cloud-ruby • google-cloud-firestore/lib/google/cloud/firestore/collection_reference_list.rb:
  120:           cols.instance_variable_set :@token, token

google-cloud-ruby • google-cloud-firestore/lib/google/cloud/firestore/document_reference/list.rb:
  174:             documents.instance_variable_set :@token, token

google-cloud-ruby • google-cloud-logging/lib/google/cloud/logging/entry/list.rb:
  156:             entries.instance_variable_set :@token,     token

google-cloud-ruby • google-cloud-logging/lib/google/cloud/logging/log/list.rb:
  137:             logs.instance_variable_set :@token,    token

google-cloud-ruby • google-cloud-logging/lib/google/cloud/logging/metric/list.rb:
  152:             metrics.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-logging/lib/google/cloud/logging/resource_descriptor/list.rb:
  158:             rds.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-logging/lib/google/cloud/logging/sink/list.rb:
  151:             sinks.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-pubsub/lib/google/cloud/pubsub/snapshot/list.rb:
  152:             subs.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-pubsub/lib/google/cloud/pubsub/subscription/list.rb:
  157:             subs.instance_variable_set :@token,   token
  172:             subs.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-pubsub/lib/google/cloud/pubsub/topic/list.rb:
  151:             topics.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-resource_manager/lib/google/cloud/resource_manager/project/list.rb:
  148:             projects.instance_variable_set :@token,   gapi_list.next_page_token

google-cloud-ruby • google-cloud-spanner/lib/google/cloud/spanner/database/list.rb:
  154:             databases.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-spanner/lib/google/cloud/spanner/instance/list.rb:
  150:             instances.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-spanner/lib/google/cloud/spanner/instance/config/list.rb:
  153:               configs.instance_variable_set :@token,   token

google-cloud-ruby • google-cloud-storage/lib/google/cloud/storage/bucket/list.rb:
  153:             buckets.instance_variable_set :@token, gapi_list.next_page_token

google-cloud-ruby • google-cloud-storage/lib/google/cloud/storage/file/list.rb:
  170:             files.instance_variable_set :@token, gapi_list.next_page_token

google-cloud-ruby • google-cloud-storage/lib/google/cloud/storage/hmac_key/list.rb:
  160:             hmac_keys.instance_variable_set :@token, gapi_list.next_page_token

```

Reply:
@dazuma Is it because with `DelegateClass` the subclass constructor must be invoked with the delegate object?

Reply:
I'm not sure why that would be. It certainly doesn't seem to be the case now (I just tested Ruby 2.4.10 and 2.7.1). In any case, something we can look at later, since it's widespread far beyond the scope of this PR.